Alerting: Deduplicate receivers during migration

codesome opened the pull request:
If 2 panel alerts have the same set of notification channels used, this PR deduplicates them and avoids creating new contact points for such channel groups.

Fixes https://github.com/grafana/grafana/issues/36697

codesome committed:
Alerting: Deduplicate receivers during migration

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

kylebrandt reviewed (approved):
Looks good to me and works with my instance. Should be easier for users and use to debug with less contact points being created <3

papagian commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -161,6 +177,25 @@ func (m *migration) makeReceiverAndRoute(ruleUid string, channelUids []interface
+		switch uid := u.(type) {
+		case string:
+			uids = append(uids, uid)
+		case int:
+			uids = append(uids, fmt.Sprintf("%d", uid))
+		default:
+			// Should never happen.
+			return "", errors.New("unknown channel UID type")
```
```suggestion
		case int, int32, int64:
			uids = append(uids, fmt.Sprintf("%d", uid))
		default:
			// Should never happen.
			return "", fmt.Errorf("unknown channel UID type: %T", u)
```

papagian commented:
I've tested it and the duplicates are reduced but I think I found another edge case that can result to duplicates:
if 2 different alerts have only [an obsolete notification channel](https://github.com/grafana/grafana/blob/ac26ca4af7a43f5417aa0e8af7e7f38dffb7e034/pkg/services/sqlstore/migrations/ualert/channel.go#L143-L145) then we will end up 2 contact point with the default channels.

codesome committed:
Fix review comments

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

codesome commented:
@papagian nice catch! Fixed it by filtering the UIDs before getting the key, and also added a log when we are dropping from UID.

codesome committed:
Better naming of contact points

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

codesome commented:
Since we are deduplicating the contact points, having rule UID in the contact point name would be confusing now and not easy to find. So I have modified the naming of contact point to be `autogen-contact-point-<number>` (also we called it recv before, which can be confusing again, changed now). It should be more friendly to find auto generated contact points now.

![Screenshot from 2021-07-29 10-58-00](https://user-images.githubusercontent.com/15064823/127436785-c0a63401-8ff7-4aa8-9e3f-f8f63dd9bb84.png)
![Screenshot from 2021-07-29 10-58-24](https://user-images.githubusercontent.com/15064823/127436795-9f3346b4-92e8-4c2c-8ead-7d312b9fde9d.png)

papagian commented:
It works much better know and I prefer the new naming a lot!
I think we could reduce further the number of policies created for 2 more corner cases (but not completely sure)
Consider the following screenshots:
<img width="1150" alt="Screenshot 2021-07-29 at 11 39 39 AM" src="https://user-images.githubusercontent.com/1632407/127460603-a8500b2d-221d-41f2-bc0b-19dff7a82863.png">
<img width="1134" alt="Screenshot 2021-07-29 at 11 39 32 AM" src="https://user-images.githubusercontent.com/1632407/127460608-34dcd95f-60e9-422e-8f76-76ee5a41054d.png">

Rules `UZgwpXW7k`, `XWgQpXZnz`, `jZgQpuW7k` are migrated dashboard alerts that referenced the webhook notifier  (using its `id`) that happened to be also one of the default notification channels.
Rule `DWgwtuZ7z` is migrated dashboard alert that references only an obsolete (non-existing) notifier (using `uid`).

Note also that `autogen-contact-point-2` contains exactly the same contact points with `autogen-contact-point-default`
```
      {
        "name": "autogen-contact-point-default",
        "grafana_managed_receiver_configs": [
          {
            "uid": "TVRwpXW7z",
            "name": "webhook",
            "type": "webhook",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "password": "",
              "severity": "critical",
              "uploadImage": true,
              "url": "http://localhost:3010",
              "username": ""
            },
            "secureFields": {}
          },
          {
            "uid": "T4RQtXZ7kz",
            "name": "#test-slack-alerts-webhook-url",
            "type": "slack",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "iconEmoji": "",
              "iconUrl": "",
              "mentionGroups": "",
              "mentionUsers": "",
              "recipient": "#test-alerts",
              "severity": "critical",
              "token": "",
              "uploadImage": true,
              "url": "",
              "username": ""
            },
            "secureFields": {
              "url": true
            }
          }
        ]
      }
```

```
{
        "name": "autogen-contact-point-2",
        "grafana_managed_receiver_configs": [
          {
            "uid": "DZgQtuZnkz",
            "name": "webhook",
            "type": "webhook",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "password": "",
              "severity": "critical",
              "uploadImage": true,
              "url": "http://localhost:3010",
              "username": ""
            },
            "secureFields": {}
          },
          {
            "uid": "vZgwpuWnzm",
            "name": "#test-slack-alerts-webhook-url",
            "type": "slack",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "iconEmoji": "",
              "iconUrl": "",
              "mentionGroups": "",
              "mentionUsers": "",
              "recipient": "#test-alerts",
              "severity": "critical",
              "token": "",
              "uploadImage": true,
              "url": "",
              "username": ""
            },
            "secureFields": {
              "url": true
            }
          }
        ]
      },
```

codesome committed:
Handle some edge cases

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

codesome committed:
Don't add routes which point to default contact point

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

papagian reviewed (approved):
I have tested and it works great!
